Userland/Libraries/LibWeb/HTML/FormAssociatedElement.cpp

```diff
@@ -589,4 +586,240 @@ void FormAssociatedTextControlElement::set_the_selection_range(Optional<WebIDL::
+    selection_was_changed();
+}
+
+void FormAssociatedTextControlElement::set_selection_focus(JS::NonnullGCPtr<DOM::Node> focus_node, size_t focus_offset)
+{
+    auto text_node = form_associated_element_to_text_node();
+    if (!text_node || focus_node != text_node)
+        return;
+    m_selection_end = focus_offset;
+    selection_was_changed();
+}
+
+void FormAssociatedTextControlElement::move_cursor_to_start(CollapseSelection collapse)
+{
+    auto text_node = form_associated_element_to_text_node();
+    if (!text_node)
+        return;
+    if (collapse == CollapseSelection::Yes) {
+        collapse_selection_to_offset(0);
+    } else {
+        m_selection_end = 0;
+    }
+    selection_was_changed();
+}
+
+void FormAssociatedTextControlElement::move_cursor_to_end(CollapseSelection collapse)
+{
+    auto text_node = form_associated_element_to_text_node();
+    if (!text_node)
+        return;
+    if (collapse == CollapseSelection::Yes) {
+        collapse_selection_to_offset(text_node->length());
+    } else {
+        m_selection_end = text_node->length();
+    }
+    selection_was_changed();
+}
+
+void FormAssociatedTextControlElement::increment_cursor_position_offset(CollapseSelection collapse)
+{
+    auto const text_node = form_associated_element_to_text_node();
+    if (!text_node)
+        return;
+    if (auto offset = text_node->grapheme_segmenter().next_boundary(m_selection_end); offset.has_value()) {
+        if (collapse == CollapseSelection::Yes) {
+            collapse_selection_to_offset(*offset);
+        } else {
+            m_selection_end = *offset;
+        }
+    }
+    selection_was_changed();
+}
+
+void FormAssociatedTextControlElement::decrement_cursor_position_offset(CollapseSelection collapse)
+{
+    auto const text_node = form_associated_element_to_text_node();
+    if (!text_node)
+        return;
+    if (auto offset = text_node->grapheme_segmenter().previous_boundary(m_selection_end); offset.has_value()) {
+        if (collapse == CollapseSelection::Yes) {
+            collapse_selection_to_offset(*offset);
+        } else {
+            m_selection_end = *offset;
+        }
+    }
+    selection_was_changed();
+}
+
+static bool should_continue_beyond_word(Utf8View const& word)
+{
+    for (auto code_point : word) {
+        if (!Unicode::code_point_has_punctuation_general_category(code_point) && !Unicode::code_point_has_separator_general_category(code_point))
+            return false;
+    }
+
+    return true;
+}
+
+void FormAssociatedTextControlElement::increment_cursor_position_to_next_word(CollapseSelection collapse)
+{
+    auto const text_node = form_associated_element_to_text_node();
+    if (!text_node)
+        return;
+
+    while (true) {
+        if (auto offset = text_node->word_segmenter().next_boundary(m_selection_end); offset.has_value()) {
+            auto word = text_node->data().code_points().substring_view(m_selection_end, *offset - m_selection_end);
```

Comment:
This one might have been an existing issue, but I don't think `code_points()` is correct here either